### PR TITLE
Enable Open Policy Agent filter in Skipper

### PR DESCRIPTION
**Pull request by mjungsbluth**
This PR introduces configuration to use Skipper with Open Policy Agent enabled as per https://github.com/zalando/skipper/pull/2364

### CODE REVIEW

File: cluster/manifests/skipper/configmap-open-policy-agent.yaml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+          s3_signing:
+            metadata_credentials:
+              aws_region: eu-central-1
+              iam_role: open-policy-agent-instance
+        name: styra-bundles
+        url: "{{ .ConfigItems.skipper_open_policy_agent_bundles_url }}"
+kind: ConfigMap
+metadata:
+  labels:
+    application: open-policy-agent-control-plane
```
Comment (mikkeloscar) [inline cluster/manifests/skipper/configmap-open-policy-agent.yaml]: Since this is the config read by skipper-ingress I think it's also fair to use `skipper-ingress` as the application label.

File: cluster/manifests/skipper/configmap-open-policy-agent.yaml [hunk 2]
```diff
@@ -0,0 +1,32 @@
+{{ if eq .ConfigItems.skipper_open_policy_agent_enabled "true" }}
```
Comment (mikkeloscar) [inline cluster/manifests/skipper/configmap-open-policy-agent.yaml:1]: Please add an entry in `deletions.yaml` to remove the config-map in case it was previously enabled but is disabled again.

See example: https://github.com/zalando-incubator/kubernetes-on-aws/blob/e85e0b03498144f81e7d6488e727435764694a25/cluster/manifests/deletions.yaml#L115-L119

File: cluster/manifests/skipper/deployment.yaml [hunk 3]
```diff
@@ -35,6 +35,9 @@ spec:
         prometheus.io/scrape: "true"
 {{- if eq .ConfigItems.skipper_topology_spread_enabled "true" }}
         zalando.org/topology-spread-timeout: 7m
+{{- end }}
+{{- if eq .ConfigItems.skipper_open_policy_agent_enabled "true" }}
+        iam.amazonaws.com/role: "{{ .ConfigItems.skipper_open_policy_agent_instance_role_arn }}"
```
Comment (mjungsbluth) [inline cluster/manifests/skipper/deployment.yaml]: This is sth that I'd like to discuss. This will be used to read bundles from S3. It would be better to use a service account IMO and assume a role bound to Skipper. Any preferences?

Comment (mikkeloscar) [reply cluster/manifests/skipper/deployment.yaml]: Yes, we should use service account (the OIDC based model) this is used in all other components in this repo. Ideally you can also provision a dedicated role as part of the cluster so it's all managed in one place. See how we do it for other roles here: https://github.com/zalando-incubator/kubernetes-on-aws/blob/d36fd80bc9a7671a9adb20dae588912baede3a45/cluster/cluster.yaml#L1520-L1654 we also have a role naming pattern including the LocalID, which you can follow as well. The idea is that we in theory can have more than one cluster per account (though we don't do it in practice at the moment).

Comment (mjungsbluth) [reply cluster/manifests/skipper/deployment.yaml]: Sounds good, I added a new role following the same pattern and assigned it to the existing `skipper-ingress` service account.

File: cluster/manifests/skipper/configmap-open-policy-agent.yaml [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  labels:
+    application: skipper-ingress
```
Comment (szuecs) [inline cluster/manifests/skipper/configmap-open-policy-agent.yaml:6]: Please add also
`component: ingress`
or can we offload this to `component: routesrv` (I guess not)?

Comment (mjungsbluth) [reply cluster/manifests/skipper/configmap-open-policy-agent.yaml:6]: Yes, changed to `component: ingress`

File: cluster/manifests/skipper/secret.yaml [hunk 5]
```diff
@@ -10,3 +10,7 @@ data:
   datadome-api-key: "{{ .Cluster.ConfigItems.datadome_api_key | base64 }}"
   lightstep-token: "{{ .Cluster.ConfigItems.lightstep_token | base64 }}"
   encryption-key: "{{ .Cluster.ConfigItems.skipper_ingress_encryption_key | base64 }}"
+  styra-token: "{{ `placeholder` | base64}}"
+#{{ if eq .Cluster.ConfigItems.skipper_open_policy_agent_enabled "true" }}
+#  styra-token: "{{ .Cluster.ConfigItems.skipper_styra_token | base64 }}"
```
Comment (szuecs) [inline cluster/manifests/skipper/secret.yaml]: commented out secret

**Comment by rishiraj88** [comment]
.

**Comment by linki** [comment]
I rebased this branch on top of `kube-1.24` branch (no conflicts). This is basically the "dev" branch until all clusters are upgraded to 1.24.

File: cluster/config-defaults.yaml [hunk 6]
```diff
@@ -947,3 +950,14 @@ sysctl_settings: ""
...
+# Set defaults values to be used by skipper ingress when the open policy agent is enabled
+{{if eq .Cluster.Environment "test"}}
+skipper_open_policy_agent_bucket_arn: "arn:aws:s3:::zalando-open-policy-agent-sandbox-bundles"
+skipper_open_policy_agent_observability_url: "https://open-policy-agent-sandbox-observability.corporate-iam.zalan.do/v1"
+skipper_open_policy_agent_bundles_url: "https://zalando-open-policy-agent-sandbox-bundles.s3.eu-central-1.amazonaws.com"
+{{else}}
+skipper_open_policy_agent_bucket_arn: "arn:aws:s3:::zalando-open-policy-agent-prod-bundles"
+skipper_open_policy_agent_observability_url: "https://open-policy-agent-observability.corporate-iam.zalan.do/v1"
+skipper_open_policy_agent_bundles_url: "https://zalando-open-policy-agent-prod-bundles.s3.eu-central-1.amazonaws.com"
+{end}}
```
Comment (linki) [inline cluster/config-defaults.yaml]: There's one `{` missing here.

File: cluster/config-defaults.yaml [hunk 7]
```diff
@@ -947,3 +950,14 @@ sysctl_settings: ""
 
 # enables/disables the minDomains field for pod topology spread.
 min_domains_in_pod_topology_spread_enabled: "true"
+
+# Set defaults values to be used by skipper ingress when the open policy agent is enabled
+{{if eq .Cluster.Environment "test"}}
```
Comment (linki) [inline cluster/config-defaults.yaml]: Let's put this block of config items next to the related one `skipper_open_policy_agent_enabled`

File: cluster/config-defaults.yaml [hunk 8]
```diff
@@ -947,3 +950,14 @@ sysctl_settings: ""
...
+skipper_open_policy_agent_bucket_arn: "arn:aws:s3:::zalando-open-policy-agent-sandbox-bundles"
+skipper_open_policy_agent_observability_url: "https://open-policy-agent-sandbox-observability.corporate-iam.zalan.do/v1"
+skipper_open_policy_agent_bundles_url: "https://zalando-open-policy-agent-sandbox-bundles.s3.eu-central-1.amazonaws.com"
+{{else}}
+skipper_open_policy_agent_bucket_arn: "arn:aws:s3:::zalando-open-policy-agent-prod-bundles"
+skipper_open_policy_agent_observability_url: "https://open-policy-agent-observability.corporate-iam.zalan.do/v1"
+skipper_open_policy_agent_bundles_url: "https://zalando-open-policy-agent-prod-bundles.s3.eu-central-1.amazonaws.com"
```
Comment (linki) [inline cluster/config-defaults.yaml]: Let's also define a "default" value for token as empty string `""`, similar to https://github.com/zalando-incubator/kubernetes-on-aws/blob/a06e8318aef6758a1d7a0987a3e8ca0b21fe7323/cluster/config-defaults.yaml#L861

This way all possible config items are documented in this file.

File: cluster/cluster.yaml [hunk 9]
```diff
@@ -1229,6 +1229,40 @@ Resources:
...
+              Federated: !Sub "arn:aws:iam::${AWS::AccountId}:oidc-provider/{{.Cluster.LocalID}}.{{ .Values.hosted_zone }}"
+            Action:
+              - 'sts:AssumeRoleWithWebIdentity'
+            Condition:
+              StringEquals:
+                "{{ .Cluster.LocalID }}.{{ .Values.hosted_zone }}:sub": "system:serviceaccount:kube-system:skipper-ingress"
+        Version: 2012-10-17
+      Path: /
+      Policies:
+        {{ if eq .Cluster.ConfigItems.skipper_open_policy_agent_enabled "true" }}
```
Comment (linki) [inline cluster/cluster.yaml]: This `if` is superfluous as it's already within another `if` block.

File: cluster/config-defaults.yaml [hunk 10]
```diff
@@ -288,6 +288,21 @@ skipper_ingress_backend_traffic_algorithm: "traffic-segment-predicate"
 # TODO: after a while we can remove this and hardcode (2023-06-30)
 skipper_ingress_default_lb_algorithm: "powerOfRandomNChoices"
 
+# Open Policy Agent
```
Comment (linki) [inline cluster/config-defaults.yaml]: Let's improve the comment of this config item. Currently it's not clear what it does.

File: cluster/config-defaults.yaml [hunk 11]
```diff
@@ -946,4 +961,4 @@ control_plane_asg_lifecycle_hook: "true"
 sysctl_settings: ""
 
 # enables/disables the minDomains field for pod topology spread.
-min_domains_in_pod_topology_spread_enabled: "true"
+min_domains_in_pod_topology_spread_enabled: "true"
```
Comment (linki) [inline cluster/config-defaults.yaml]: This newline change looks unrelated. Let's not change it here.

File: cluster/manifests/skipper/configmap-open-policy-agent.yaml [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+      resource: v1/bundles/zalando-apps/{{ `{{ .bundlename }}` }}/discovery.tgz
+      service: styra-bundles
+    labels:
+      environment: {{ .Cluster.Environment }}
+      kubernetes-cluster-alias: {{ .Cluster.Alias }}
+      aws-account-id: "{{ accountID .Cluster.InfrastructureAccount }}"
+    services:
+    - credentials:
+        bearer:
+          token: "{{ `{{ .Env.STYRA_TOKEN }}` }}"
```
Comment (linki) [inline cluster/manifests/skipper/configmap-open-policy-agent.yaml:23]: Since this is a secret value it should ideally be stored in a `Secret` object and not it a `ConfigMap`.

File: cluster/manifests/skipper/configmap-open-policy-agent.yaml [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+      name: styra
+      url: "{{ .ConfigItems.skipper_open_policy_agent_observability_url }}"
+    - credentials:
+        s3_signing:
+          web_identity_credentials:
+            aws_region: eu-central-1
```
Comment (linki) [inline cluster/manifests/skipper/configmap-open-policy-agent.yaml:29]: Let's use `{{ .Cluster.Region }}` here.

Comment (linki) [reply cluster/manifests/skipper/configmap-open-policy-agent.yaml:29]: I'm assuming that `eu-west-1` would be the appropriate value if the cluster was in the `eu-west-1` region. If that's not the case, please disregard my comment.

File: cluster/manifests/skipper/secret.yaml [hunk 14]
```diff
@@ -10,3 +10,6 @@ data:
   datadome-api-key: "{{ .Cluster.ConfigItems.datadome_api_key | base64 }}"
   lightstep-token: "{{ .Cluster.ConfigItems.lightstep_token | base64 }}"
   encryption-key: "{{ .Cluster.ConfigItems.skipper_ingress_encryption_key | base64 }}"
+{{ if eq .Cluster.ConfigItems.skipper_open_policy_agent_enabled "true" }}
```
Comment (szuecs) [inline cluster/manifests/skipper/secret.yaml]: we can also drop the if, similar to other keys.

File: cluster/manifests/skipper/configmap-open-policy-agent.yaml (hunk 13 above)
Comment (linki) [reply cluster/manifests/skipper/configmap-open-policy-agent.yaml:29]: We decided to skip this for now because it's not clear whether this has to be the same region as the cluster all the time.

**Comment by linki** [comment]
👍

**Comment by adetunjiakintundeakinde** [comment]
👍

**Comment by demonCoder95** [comment]
👍
